src/map.cpp

```diff
@@ -3017,315 +3017,300 @@ ter_id map::get_roof( const tripoint &p, const bool allow_air )
-            bool collapses = has_flag("COLLAPSES", p) && smash_ter;
-            bool supports = has_flag("SUPPORTS_ROOF", p) && smash_ter;
-            if( smash_furn ) {
-                furn_set(p, bash->furn_set);
-                // Hack alert.
-                // Signs have cosmetics associated with them on the submap since
-                // furniture can't store dynamic data to disk. To prevent writing
-                // mysteriously appearing for a sign later built here, remove the
-                // writing from the submap.
-                delete_signage( p );
-            } else if( !smash_ter ) {
-                // Handle error earlier so that we can assume smash_ter is true below
-                debugmsg( "data/json/terrain.json does not have %s.bash.ter_set set!",
-                          ter_at(p).id.c_str() );
-            } else if( bash->ter_set != null_ter_t ) {
-                // If the terrain has a valid post-destroy terrain, set it
-                ter_set( p, bash->ter_set );
-            } else {
-                tripoint below( p.x, p.y, p.z - 1 );
-                const auto &ter_below = ter_at( below );
-                if( bash->bash_below && ter_below.has_flag( "SUPPORTS_ROOF" ) ) {
-                    // When bashing the tile below, don't allow bashing the floor
-                    bash_ter_furn( below, str, silent, destroy, false, res_roll );
+                if ( bash->str_max_supported != -1 ) {
+                    smax = bash->str_max_supported;
                 }
+            }
+        }
+        // Linear interpolation from str_min to str_max
+        const int resistance = smin + ( params.roll * ( smax - smin ) );
+        if( params.strength >= resistance ) {
+            success = true;
+        }
+    }
+
+    if( !params.destroy && !success ) {
+        if( sound_fail_vol == -1 ) {
+            sound_volume = 12;
+        } else {
+            sound_volume = sound_fail_vol;
+        }
+
+        sound = _(bash->sound_fail.c_str());
+        params.did_bash = true;
+        if( !params.silent ) {
+            sounds::sound( p, sound_volume, sound, false, "bash", sound );
+        }
+
+        return;
+    }
+
+    // Clear out any partially grown seeds
+    if( has_flag_ter_or_furn( "PLANT", p ) ) {
+        i_clear( p );
+    }
+
+    if( ( smash_furn && has_flag_furn("FUNGUS", p) ) ||
+        ( smash_ter && has_flag_ter("FUNGUS", p) ) ) {
+        create_spores( p );
+    }
 
-                ter_set( p, t_open_air );
+    if( params.destroy ) {
+        sound_volume = smax;
+    } else {
+        if( sound_vol == -1 ) {
+            sound_volume = std::min(int(smin * 1.5), smax);
+        } else {
+            sound_volume = sound_vol;
+        }
+    }
+
+    sound = _(bash->sound.c_str());
+    // Set this now in case the ter_set below changes this
+    const bool collapses = has_flag("COLLAPSES", p) && smash_ter;
+    const bool supports = has_flag("SUPPORTS_ROOF", p) && smash_ter;
```

Comment:
If you want better performance, the boolean variable should be the first and the function the second operand. If the variable already contains `false`, the function won't be called at all.